File: orderset.cpp
#include "bits/stdc++.h"
using namespace std;
struct orderset {
	struct node {
		node *l, *r;
		int value;
		int prio;

		int size;
	};
	using pnode = node*;
	pnode create(int value) {
		pnode p = (pnode) malloc(sizeof (node));
		p -> value = value;
		p -> prio = rand(); // One have to ensure that all the priorities are disinct
		p -> l = nullptr;
		p -> r = nullptr;

		p -> size = 1;
		return p;
	}

	int size(pnode t) {
		return t ? t -> size : 0;
	}
	void update(pnode &t) {
		if(t) {
			t -> size = size(t -> l) + size(t -> r) + 1;
		}
	}

	void split(pnode t, int key, pnode& a, pnode &b) {
		if(t == nullptr) {
			a = b = nullptr;
			return ;
		}
		if(t -> value <= key) {
			a = t;
			split(t -> r, key, a -> r, b);
		} else {
			b = t;
			split(t -> l, key, a, b -> l);
		}
		update(a);
		update(b);
	}
	pnode merge(pnode a, pnode b) {
		if(a == nullptr) return b;
		if(b == nullptr) return a;
		
		if(a -> prio < b -> prio) {
			a -> r = merge(a -> r, b);
			update(a);
			return a;
		} else {
			b -> l = merge(a, b -> l);
			update(b);
			return b;
		}
	}

	void add(pnode &t, pnode a) {
		if(t == nullptr) {
			t = a;
			return ;
		}	 
		if(t -> prio < a -> prio) {
			if(a -> value <= t -> value) {
				add(t -> l, a);
			} else {
				add(t -> r, a);
			}
		} else {
			split(t, a -> value, a -> l, a -> r);		
			t = a;
		}
		update(t);
	}
	void del(pnode &t, int value) {
		if(t == nullptr) return ;
		if(t -> value == value) {
	 		pnode temp = t;
	 		t = merge(t -> l, t -> r);
	 		free(temp);
		} else if (value <= t -> value) {
			del(t -> l, value);
		} else {
			del(t -> r, value);
		}
		update(t);
	}
	pnode kth(pnode t, int k) {
		if(t == nullptr) return t;
		if(size(t -> l) + 1 == k) return t;
		else if (size(t -> l) >= k) return kth(t -> l, k);
		else return kth(t -> r, k - size(t -> l) - 1); 
	}

	int order(pnode t, int value) {
		if(t == nullptr) return 0;
		if(t -> value == value) {
			return size(t -> l) + 1;
		} else if (value <= t -> value) {
			return order(t -> l, value);
		} else {
			return order(t -> r, value) + size(t -> l) + 1;
		}
	}
	void print(pnode t) {
		if(t == nullptr) return ;
		print(t -> l);
		cerr << t->value << endl;
		print(t -> r);
	}

	pnode t;
	orderset () {
		t = nullptr;
	}
	void insert(int value) {
		add(t, create(value));
	}
	void erase(int value) {
		del(t, value);
	}
	int at(int idx) {
		return kth(t, idx) -> value;
	}
	int find(int value) {
		return order(t, value);
	}
	bool present(int value) {
		int res = find(value);
		if(res > size() or res == 0) return false;
		return res == at(res);
	}
	int size() {
		return size(t);
	}
};

int main(int argc, char const *argv[])
{
	orderset s;
	s.insert(12);
	s.insert(34);
	s.insert(51);
	s.insert(7);
	s.insert(24);

	cerr << "Elements are:" << endl; 
	for(int i = 1; i <= s.size(); i++) {
		cerr << s.at(i) << " ";
	}
	cerr << endl;
	
	cerr << s.find(7) << endl;
	cerr << s.find(100) << endl;
	cerr << s.find(3) << endl;
	cerr << s.find(34) << endl;
	return 0;
}

File: Orderset.java
import java.util.*;
import java.io.*;

class OrderSet {
	private class Node {
		int l, r;
		int sum;
		Node () {
			sum = 0;
			l = -1;
			r = -1;
		}
	}
	private ArrayList <Node> t;
	private int maxn;
	private int now;
	
	OrderSet (int maxn) {
		this.maxn = maxn;
		t = new ArrayList <> ();
		t.add(new Node());
		now = 0;
	}
	private void update(int x, int c, int b, int e) {		
		Node p = t.get(c);
		if(b == e) {
			p.sum += 1;
			return ;
		}
		int m = (b + e) >> 1;
		if(x <= m) {
			if(p.l == -1) {
				p.l = ++now;
				t.add(new Node());
			}
			update(x, p.l, b, m);
		}
		else {
			if(p.r == -1) {
				p.r = ++now;
				t.add(new Node());
			}
			update(x, p.r, m+1, e);
		}
		p.sum = 0;
		if(p.l != -1) p.sum += t.get(p.l).sum;
		if(p.r != -1) p.sum += t.get(p.r).sum; 
	}
	private int range(int x, int y, int c, int b, int e) {
		if(c == -1) return 0;
		if(x <= b && e <= y) return t.get(c).sum;
		int m = (b + e) >> 1;
		int l = t.get(c).l;
		int r = t.get(c).r;
		return range(x, y, l, b, m) + range(x, y, r, m+1, e);
	}
	private int find(int x, int c, int b, int e) {
		if(c == -1) return -1;
		if(b == e) {
			return b;
		}
		int m = (b + e) >> 1;
		int l = t.get(c).l;
		int r = t.get(c).r;
		if(l != -1 && x <= t.get(l).sum) {
			return find(x, l, b, m);
		} else {
			if(l == -1) return find(x, r, m+1, e);
			else return find(x - t.get(l).sum, r, m+1, e);
		}
	}
	public void add(int x) {
		update(x, 0, 0, maxn);
	}
	public int get(int x) {
		return find(x+1, 0, 0, maxn);
	}
	public int count(int l, int r) {
		return range(l, r, 0, 0, maxn);
	} 
	public int size() {
		return t.get(0).sum;
	}
}

public class Orderset {
	public static void main(String[] Args) {
		OrderSet t = new OrderSet (1000);
		t.add(3);
		t.add(4);
		t.add(6);
		t.add(11);
		t.add(7);
		t.add(0);
		t.add(89);
		t.add(7);
		
		for(int i = 0; i < t.size(); i++) {
			System.out.println(t.get(i));
		}
		System.out.println(t.count(6, 260));
	}
}
